drivers/serial/uart_intel_lw.c

```diff
@@ -0,0 +1,999 @@
+#define INTEL_LW_UART_CONTROL_RRDY_MSK     (0x80)
+#define INTEL_LW_UART_CONTROL_E_MSK        (0x100)
+#define INTEL_LW_UART_CONTROL_TRBK_MSK     (0x200)
+#define INTEL_LW_UART_CONTROL_DCTS_MSK     (0x400)
+#define INTEL_LW_UART_CONTROL_RTS_MSK      (0x800)
+#define INTEL_LW_UART_CONTROL_EOP_MSK      (0x1000)
+
+/* defined values */
+#define UART_INTEL_LW_NO_ERROR (0u)
+#define INTEL_LW_UART_CLEAR_STATUS_VAL (0u)
+#define INTEL_LW_UART_PENDING_MASK  (INTEL_LW_UART_STATUS_RRDY_MSK | \
+			INTEL_LW_UART_STATUS_TRDY_MSK | INTEL_LW_UART_STATUS_E_MSK \
+			| INTEL_LW_UART_STATUS_EOP_MSK)
+
+/***********************/
+/* configuration flags */
+/*
+ * The value INTEL_LW_UART_FB is a value set in the devices flag field to
+ * indicate that the device has a fixed baud rate; i.e. if this flag is set
+ * software can not control the baud rate of the device.
+ */
+#define INTEL_LW_UART_FB 0x1
+
+/*
+ * The value INTEL_LW_UART_FC is a value set in the device flag field to
+ * indicate the the device is using flow control, i.e. the driver must
+ * throttle on transmit if the nCTS pin is low.
+ */
+#define INTEL_LW_UART_FC 0x2
+
+/* end of configuration flags */
+/******************************/
+
+/* device data */
+struct uart_intel_lw_device_data {
+	struct uart_config uart_cfg; /* stores uart config from device tree*/
+	struct k_spinlock lock;
+	uint32_t status_act; /* stores value of status register. */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t cb;  /**< Callback function pointer */
+	void *cb_data;  /**< Callback function arg */
+#ifdef CONFIG_UART_INTEL_LW_EOP
+	uint8_t set_eop_cb;
+	uart_irq_callback_user_data_t cb_eop;  /**< Callback function pointer */
+	void *cb_data_eop;  /**< Callback function arg */
+#endif /* CONFIG_UART_INTEL_LW_EOP */
+	uint32_t control_val; /* stores value to set control register. */
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+/*
+ * device config:
+ * stores data that cannot be changed during run time.
+ */
+struct uart_intel_lw_device_config {
+	mm_reg_t base;
+	uint32_t flags;              /* refer to configuration flags */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t  irq_config_func;
+	unsigned int irq_num;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+/**
+ * function prototypes
+ */
+static int uart_intel_lw_irq_update(const struct device *dev);
+static int uart_intel_lw_irq_tx_ready(const struct device *dev);
+static int uart_intel_lw_irq_rx_ready(const struct device *dev);
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+/**
+ * @brief Poll the device for input.
+ *
+ * This is a non-blocking function.
+ *
+ * This driver supports interrupt driven api.
+ * Polling for data under normal operation, might cause unexpected behaviour.
+ * If users wish to poll for data, please use the api:
+ * uart_drv_cmd with CMD_POLL_ASSERT_RTS before polling in.
+ * Then use CMD_POLL_DEASSERT_RTS to resume normal operation after polling.
```

Comment:
This is going to be problematic if someone is running some uart tests in the tree. Those tests do not assert RTS before calling `uart_poll_in()`. Besides, shouldn't this be asserting CTS instead to indicate to other side they can send something.

Reply:
The reason for manual RTS assertion is for efficiency.

For example, if user would like to poll out 10 characters.
It would be better for user to assert/deassert RTS once in the application instead of doing it 10 times in the poll_out.
I have also added CONFIG_UART_INTEL_LW_AUTO_LINE_CTRL_POLL as a choice.

For poll in, as long as the user did not disable the HW interrupt, the driver should be able to handle the flow control via uart_intel_lw_dcts_isr.

I have modified the description too. Please review again. Thank you.